File: src/main/kotlin/demo/kotliln/boardgame/service/rating/BoardGameRatingService.java
package demo.kotliln.boardgame.service.rating;

import demo.kotliln.boardgame.api.models.AgeRange;
import demo.kotliln.boardgame.api.models.BoardGame;
import demo.kotliln.boardgame.api.models.Category;
import demo.kotliln.boardgame.api.models.NumberOfPlayers;
import demo.kotliln.boardgame.api.models.Rating;
import demo.kotliln.boardgame.service.boardgame.BoardGameRepository;
import demo.kotliln.boardgame.service.entities.BoardGameEntity;
import demo.kotliln.boardgame.service.exceptions.BoardGameNotFound;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
public class BoardGameRatingService {
    private final BoardGameRepository boardGameRepository;
    private final RateRepository rateRepository;

    @Autowired
    public BoardGameRatingService(final BoardGameRepository boardGameRepository, final RateRepository rateRepository) {
        this.boardGameRepository = boardGameRepository;
        this.rateRepository = rateRepository;
    }

    public BoardGame ratingBoardGame(final RateBoardGame rateBoardGame) throws BoardGameNotFound {
        final RateEntity rate = new RateEntity(null,
                rateBoardGame.getBoardGameName(),
                rateBoardGame.getRate()
        );

        log.debug("Saving new rating entity for boardgame {}",
                rateBoardGame.getBoardGameName()
        );

        rateRepository.save(rate);

        log.debug("Return updated rate for boardgame {}",
                rateBoardGame.getBoardGameName()
        );

        return boardGameRepository.findByName(rateBoardGame.getBoardGameName())
                .map(this::ratedBoardGame)
                .orElseThrow(
                        () -> new BoardGameNotFound(rateBoardGame.getBoardGameName())
                );
    }

    public List<BoardGame> withHigherRateThan(final Double rate) {
        log.debug("Find all boardgames with rate higher than {}", rate);

        return boardGameRepository.findAll()
                .stream()
                .map(this::ratedBoardGame)
                .filter(game -> game.getRating().getAverageRate() > rate)
                .collect(Collectors.toList());
    }

    private BoardGame ratedBoardGame(final BoardGameEntity bg) {
        log.debug("Retrieving all the rates of boardgame {}", bg.getName());

        final List<RateEntity> ratesOfTheBoardGame = rateRepository.findByBoardGameName(bg.getName());
        final Double average = ratesOfTheBoardGame.stream()
                .mapToDouble(RateEntity::getRate)
                .average()
                .getAsDouble();

        return new BoardGame(bg.getName(), Category.valueOf(bg.getCategory().name()),
                new Rating(average),
                new AgeRange(bg.getMinimalAge(), bg.getMaximalAge()),
                new NumberOfPlayers(bg.getMinimalNumberOfPlayers(), bg.getMaximalNumberOfPlayers())
        );
    }
}

File: src/main/kotlin/demo/kotliln/boardgame/service/rating/BoardGameRatingService.kt
package demo.kotliln.boardgame.service.rating

import demo.kotliln.boardgame.api.models.*
import demo.kotliln.boardgame.service.boardgame.entities.BoardGameEntity
import demo.kotliln.boardgame.service.boardgame.exceptions.BoardGameNotFound
import demo.kotliln.boardgame.service.boardgame.repositories.BoardGameRepository
import demo.kotliln.boardgame.service.rating.entities.RateBoardGame
import demo.kotliln.boardgame.service.rating.entities.RateEntity
import demo.kotliln.boardgame.service.rating.repositories.RateRepository
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
open class BoardGameRatingService (
        private val boardGameRepository: BoardGameRepository,
        private val rateRepository: RateRepository
) {
    private val logger = KotlinLogging.logger { }

    fun ratingBoardGame(rateBoardGame: RateBoardGame): BoardGame {
        val rateEntity = RateEntity(null, rateBoardGame.boardGameName, rateBoardGame.rate)

        logger.debug { "Saving new rating entity for boardgame ${rateBoardGame.boardGameName}" }
        rateRepository.save(rateEntity)
        logger.debug { "Return updated rate for boardgame ${rateBoardGame.boardGameName}" }

        return boardGameRepository.findByName(rateBoardGame.boardGameName)
                .map { this.ratedBoardGame(it) }
                .orElseThrow { BoardGameNotFound(rateBoardGame.boardGameName) }
    }

    fun withHigherRateThan(rate: Double): List<BoardGame> {
        logger.debug("Find all boardgames with rate higher than $rate")

        return boardGameRepository.findAll()
                .map{ this.ratedBoardGame(it) }
                .filter { it.rating.averageRate > rate }
    }

    private fun ratedBoardGame(bg: BoardGameEntity): BoardGame {
        logger.debug("Retrieving all the rates of boardgame ${bg.name}")

        val ratesOfTheBoardGame = rateRepository.findByBoardGameName(bg.name)
        val average = ratesOfTheBoardGame.asSequence().map { it.rate }.average()

        return BoardGame(
                bg.name,
                Category.valueOf(bg.category.name),
                Rating(average),
                AgeRange(bg.minimalAge, bg.maximalAge),
                NumberOfPlayers(bg.minimalNumberOfPlayers, bg.maximalNumberOfPlayers),
        )
    }


}